garage/tf/models/base.py

```diff
@@ -117,32 +104,75 @@ def outputs(self):
 
     def __getstate__(self):
         args, kwargs = self.init_spec
-        return {
-            'args': args,
-            'kwargs': kwargs,
-            'weights': self.weights
-        }
+        return {'args': args, 'kwargs': kwargs, 'parameters': self.parameters}
 
     def __setstate__(self, data):
         # TODO(krzentner): There has got to be a better way.
         out = type(self)(*data['args'], **data['kwargs'])
-        out._default_weights = data['weights']
+        out._default_parameters = data['parameters']
         self.__dict__.update(out.__dict__)
 
 
 class TfModel(Model):
-
     def __init__(self, name):
         super().__init__()
         # All TF models need a valid unique name
         # TODO(ahtsan): produce a message on collision?
         self._name = name or self._name
+        self._is_parent = True
+        self._is_built = False
         # All ops and variables for this Model *MUST* be placed under this scope
-        self._scope = tf.variable_scope(self.name, reuse=False)
+        self._variable_scope = tf.variable_scope(self.name, reuse=False)
+        self._name_scope = tf.name_scope("default")
+        self._siblings = {}
+
+    @classmethod
+    def _new_child(cls, other, name):
+        twin = cls(**other.init_spec[1])
+        twin._name = name
+        twin._is_parent = False
+        twin._default_parameters = other._default_parameters
+        twin._is_built = other._is_built
+        twin._variable_scope = other.variable_scope
+        twin._name_scope = tf.name_scope(name)
+        return twin
+
+    def build(self, *inputs, name=None):
+        if not name:
+            if 'self' in self._siblings:
+                raise ValueError("Subgraph {} already exists!".format(
+                    self.name))
+            with self._name_scope:
+                _variable_scope = self.variable_scope if not self._is_built else tf.variable_scope(
+                    self._variable_scope._name_or_scope, reuse=True)
+                with _variable_scope:
+                    self._inputs = inputs
+                    outputs = self._build(*inputs)
+                    self._outputs = list(outputs)
+                    self._siblings['self'] = self
+                    if not self._is_built:
+                        variables = self._get_variables().values()
+                        tf.get_default_session().run(
+                            tf.variables_initializer(variables))
+                        if self._default_parameters:
+                            self.parameters = self._default_parameters
+                        self._is_built = True
+        elif name in self._siblings:
+            raise ValueError("Subgraph {} already exists!".format(name))
+        elif not self._is_parent:
+            raise ValueError("Child of a model should not be built!")
+        else:
+            twin = type(self)._new_child(self, name)
+            setattr(self, name, twin)
```

Comment:
Is it a good idea to use setattr here?

Reply:
@krzentner would like your thoughts

Reply:
I don't really see why we would want to use `setattr()` here, since we could already access the graph by name using `_siblings`. If we really are going with a design like this, we should just rename `_siblings` to `instances`. Using `setattr()` looks like it's just asking for trouble down the road.